File: src/types/telop.ts
export interface TelopTemplate {
  id: string;
  name: string;
  description: string;
  category: string;
  animation: TelopAnimationDef;
  default_style: TelopStyle;
  preview_text?: string;
}

export interface TelopAnimationDef {
  unit: "character" | "word";
  property?: string;
  properties?: string[];
  from?: Record<string, number>;
  to?: Record<string, number>;
  keyframes?: Keyframe[];
  duration_ms: number;
  delay_per_unit_ms: number;
  easing: string;
}

export interface Keyframe {
  t: number;
  opacity?: number;
  translate_y?: number;
  translate_x?: number;
  scale?: number;
  rotate?: number;
}

export interface TelopStyle {
  font_family: string;
  font_size: number;
  font_weight: number;
  color: string;
  outline?: OutlineStyle;
  shadow?: ShadowStyle;
  position?: { x: number; y: number };
  alignment?: string;
}

export interface OutlineStyle {
  enabled: boolean;
  color: string;
  width: number;
  join?: "miter" | "round" | "bevel";
}

export interface ShadowStyle {
  enabled: boolean;
  color: string;
  offset_x: number;
  offset_y: number;
  blur: number;
}

File: src-tauri/src/models/telop.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SubtitleCue {
    pub index: usize,
    pub start_ms: u64,
    pub end_ms: u64,
    pub text: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TelopStyle {
    pub font_family: String,
    pub font_size: f32,
    pub font_weight: u32,
    pub color: String,
    #[serde(default)]
    pub outline: Option<OutlineStyle>,
    #[serde(default)]
    pub shadow: Option<ShadowStyle>,
    #[serde(default)]
    pub position: Option<Position>,
    #[serde(default)]
    pub alignment: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OutlineStyle {
    pub enabled: bool,
    pub color: String,
    pub width: f32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ShadowStyle {
    pub enabled: bool,
    pub color: String,
    pub offset_x: f32,
    pub offset_y: f32,
    pub blur: f32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Position {
    pub x: f32,
    pub y: f32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TelopAnimation {
    #[serde(rename = "type")]
    pub animation_type: String,
    #[serde(default)]
    pub speed: Option<String>,
    #[serde(default)]
    pub delay_per_char_ms: Option<u64>,
    #[serde(default)]
    pub easing: Option<String>,
}
